### PR TITLE
feat(integration:tailwind): add support for TS tailwind config

### LINKED ISSUE #6695: @astrojs/tailwind ignores Tailwind CSS v3.3 TypeScript config
### What version of `astro` are you using?

2.1.8

### Are you using an SSR adapter? If so, which one?

None

### What package manager are you using?

pnpm

### Describe the Bug

When I use Tailwind CSS v3.3 TypeScript config file `tailwind.config.ts`, Astro ignores it and uses default config instead.

**Reference:** https://tailwindcss.com/blog/tailwindcss-v3-3#esm-and-typescript-support

I think you just need to bump Tailwind dependency to use updated `resolveConfig`.

### Participation

- [ ] I am willing to submit a pull request for this issue.

### PR DESCRIPTION
**Pull request by FinnDore**
## Changes

This resolves #6695 by adding support for tailwind configs

## Testing

Ive tested this with the example repo in the project. WIll test a little more later. Are there any non manual test for this?

## Docs

Dont think any docs are needed, you can use ts configs now but thats it

### CODE REVIEW

**Comment by FinnDore** [comment]
seems like if i nest the tailwind config and point to that in the astro config it breaks. need to fix that

**Comment by FinnDore** [comment]
On second thaughts are nested configs even supported by tailwind and or astro? this error seems to come from the TW config loader
```
Error: Cannot find module '/Users/finn/Documents/GitHub/astro/examples/with-tailwindcss/nested/tailwind.config.ts'
Require stack:
- /Users/finn/Documents/GitHub/astro/node_modules/.pnpm/tailwindcss@3.3.0_postcss@8.4.21/node_modules/tailwindcss/lib/lib/load-config.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1039:15)
    at Function.resolve (node:internal/modules/cjs/helpers:109:19)
    at _resolve (/Users/finn/Documents/GitHub/astro/node_modules/.pnpm/jiti@1.18.2/node_modules/jiti/dist/jiti.js:1:241025)
    at jiti (/Users/finn/Documents/GitHub/astro/node_modules/.pnpm/jiti@1.18.2/node_modules/jiti/dist/jiti.js:1:243309)
    at loadConfig (/Users/finn/Documents/GitHub/astro/node_modules/.pnpm/tailwindcss@3.3.0_postcss@8.4.21/node_modules/tailwindcss/lib/lib/load-config.js:34:26)
    at getUserConfig (file:///Users/finn/Documents/GitHub/astro/packages/integrations/tailwind/dist/index.js:55:24)
    at astro:config:setup (file:///Users/finn/Documents/GitHub/astro/packages/integrations/tailwind/dist/index.js:105:34)
    at runHookConfigSetup (file:///Users/finn/Documents/GitHub/astro/packages/astro/dist/integrations/index.js:79:60)
    at async createContainer (file:///Users/finn/Documents/GitHub/astro/packages/astro/dist/core/dev/container.js:31:14)
    at async createContainerWithAutomaticRestart (file:///Users/finn/Documents/GitHub/astro/packages/astro/dist/core/dev/restart.js:90:28) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/Users/finn/Documents/GitHub/astro/node_modules/.pnpm/tailwindcss@3.3.0_postcss@8.4.21/node_modules/tailwindcss/lib/lib/load-config.js'
  ]
}
./nested/tailwind.config.ts {
  config: undefined,
  configPath: '/Users/finn/Documents/GitHub/astro/examples/with-tailwindcss/nested/tailwind.config.ts'
}
```

**Comment by FinnDore** [comment]
Ive updated the peer dep. From what i understand this will force users of the tailwind integration to use tailwind `^3.3.0`? 
ive done this beacuse prior to tailwind `3.3.0` the `loadConfig` function did not exist. (https://github.com/tailwindlabs/prettier-plugin-tailwindcss/blob/70ea7aa55d5ecc1b9202273b4c4deb84c4d8c979/src/index.js#L174) is this what we want here? honestly i dont know.

**Comment by FinnDore** [comment]
The cli keeps OOM for me. no idea why these changes seem trivial

**Comment by FinnDore** [comment]
Im leaning into the idea that the consumers should not need `3.3.0` but i dont understand how this is meant to magically load the loader if the user does not have `3.3.0`
```
// Prior to `tailwindcss@3.3.0` this won't exist so we load it last
require(path.join(pkgDir, 'loadConfig'))
```

File: packages/integrations/tailwind/package.json [hunk 1]
```diff
@@ -28,20 +28,19 @@
...
-    "@proload/core": "^0.3.2",
     "autoprefixer": "^10.4.7",
     "postcss": "^8.4.14",
     "postcss-load-config": "^4.0.1"
   },
   "devDependencies": {
     "astro": "workspace:*",
     "astro-scripts": "workspace:*",
-    "tailwindcss": "^3.0.24",
+    "tailwindcss": "3.3.0",
```
Comment (bai) [inline packages/integrations/tailwind/package.json:38]:
```suggestion
    "tailwindcss": "^3.3.0",
```

Comment (FinnDore) [reply packages/integrations/tailwind/package.json]: what was the reason for not pinning this?

Comment (FinnDore) [reply packages/integrations/tailwind/package.json]: makes sense thanks for clarifying.

**Comment by FinnDore** [comment]
it builds locally 🙈
image

**Comment by FinnDore** [comment]
even downgrading to node 16 locally and it still build fine

**Comment by FinnDore** [comment]
Yeah this is really bugging me. Is it trying to inline the import or somthing?

**Comment by FinnDore** [comment]
its OOM 😭

File: packages/integrations/tailwind/src/index.ts [hunk 2]
```diff
@@ -28,42 +33,49 @@ async function getUserConfig(root: URL, configPath?: string, isRestart = false)
...
+
+	const configPathToUse = userConfigPath ?? resolvedConfigPath;
+
+	if (!loadConfig && configPathToUse.endsWith('ts')) {
+		throw new Error(
+			'To use a typescript tailwind config file please install tailwind 3.3.0 or higher'
+		);
+	}
+
+	// previous to tailwindcss 3.3.0 loadConfig did not exist
```
Comment (juliusmarminge) [inline packages/integrations/tailwind/src/index.ts:59]:
```suggestion
	if (!loadConfig) {
	  if (configPathToUse.endsWith('ts')) {
		  throw new Error('TypeScript Tailwind configurations requires `tailwindcss@^3.3.0`');
	  }
	  if (configPathToUse.endsWith('mjs')) {
		  throw new Error('ESM Tailwind configurations requires `tailwindcss@^3.3.0`');
	  }
```

File: packages/integrations/tailwind/src/index.ts [hunk 3]
```diff
@@ -28,42 +33,49 @@ async function getUserConfig(root: URL, configPath?: string, isRestart = false)
...
-			filePath: userConfigPath,
-		})) as string;
-
-		const { dir, base } = path.parse(resolvedConfigPath);
-		const tempConfigPath = path.join(dir, `.temp.${Date.now()}.${base}`);
-		await fs.copyFile(resolvedConfigPath, tempConfigPath);
-
-		let result: load.Config<Record<any, any>> | undefined;
+		clearModule(configPathToUse);
+		if (configPathToUse.endsWith('ts')) {
```
Comment (juliusmarminge) [inline packages/integrations/tailwind/src/index.ts:63]:
```suggestion
		if (configPathToUse.endsWith('ts') || configPathToUse.endsWith("mjs") {
```

File: packages/integrations/tailwind/src/index.ts [hunk 4]
```diff
@@ -28,42 +33,49 @@ async function getUserConfig(root: URL, configPath?: string, isRestart = false)
...
+			return { config: (await import(configPathToUse))?.default, configPath: configPathToUse };
+		}
+	} else {
 		try {
-			result = await load('tailwind', {
-				mustExist: false,
-				cwd: resolvedRoot,
-				filePath: tempConfigPath,
-			});
+			if (configPathToUse.endsWith('ts')) {
```
Comment (juliusmarminge) [inline packages/integrations/tailwind/src/index.ts:70]:
```suggestion
			if (configPathToUse.endsWith('ts') || configPathToUse.endsWith('mjs')) {
```

prob time to pull this out to an `isESM` variable now xdd

**Comment by juliusmarminge** [comment]
Should probably also check if the user has `type: "module"` set and use ESM for `.js` in that case too

**Comment by FinnDore** [comment]
You did mention these before. ill commit these.

**Comment by FinnDore** [comment]
image
Well all the tailwind tests pass, however some other test fails because it looks at the root of test dir instead of the respective project for a tailwind config file. So if someone can fix my code so that tests passes i think this would be done. For now im going to close this pr.

**Comment by filipesmedeiros** [comment]
Why did you close the PR @FinnDore? Seems to make sense to leave it open, since it wasn't wrong, but also wasn't merged?

**Comment by FinnDore** [comment]
Some of the tests were failing, and only maintainers can push to this branch so there's not much use in leaving this pr open unless I'm mistaken.

It's p much there though if someone wants to fork my astro fork.

**Pull request closed without merging**